### ISSUE TITLE
Extract Action Notifier* framework for push notifications

### ISSUE THREAD

**Issue by dhh** [rails/rails#50454]
One of the most exciting developments in web standards this year was the full adoption of Web Push by Apple. This means we now have a reliable push notification standard that works in Safari, Chrome, and Firefox, which [reduces the need for native applications](https://world.hey.com/dhh/native-mobile-apps-are-optional-for-b2b-startups-in-2024-4c870d3e). But getting the service worker, web push configuration, and permission flow right is surprisingly complicated. We already have some great gems, like web-push, to handle some of the underlying plumbing, but putting it all together is far more complicated than it ought to be. So let's compress that complexity with a new dedicated framework.

While compressing the complexity of Web Push is the primary motivation for Action Notifier, it's also a great time to add an plugin setup that allows us to plug in adapters for services like pinpoint or other native platform protocols. But the intention is not to support every possible delivery format for notifications under the sun (like SMS and definitely not email, we have Action Mailer for that).

The first stab at this will be an extraction from ONCE #1. So until there's a first gem for this, let's hold any more detailed discussion of features or capabilities.

Prior art that'll be examined: https://github.com/palkan/active_delivery + ?

* The name Action Notifier is currently reserved by Magnus Hult for a gem from back in 2018. If it's not possible to take over that name, we'd have to go with something else.

**Comment by joemasilotti** [comment]
Very excited for this! For more art to be examined, I've used [Noticed](https://github.com/excid3/noticed) from @excid3 on a bunch of my projects. It's been especially helpful for wrangling iOS + Android push notifications.

**Comment by dhh** [comment]
Action Notifier should also address the need for controlling the PWA badge counter. That's tied to the hip with Web Push, since the only time you get to reliably move the counter is when a notification is delivered.

**Comment by segheysens** [comment]
Re: gem naming conflict - perhaps Action Alert or Action Broadcast?

**Comment by justinko** [comment]
Magnus Hult's agent checking in here — opening bid is $5,000

**Comment by manuelvanrijn** [comment]
> Re: gem naming conflict - perhaps Action Alert or Action Broadcast?

Or maybe Action Notify or Action Notification. Broadcast might be a bit confusing because of ActionCable?

**Comment by yegorch** [comment]
Very interesting)

**Comment by Natblow** [comment]
> > Re: gem naming conflict - perhaps Action Alert or Action Broadcast?
> 
> Or maybe Action Notify or Action Notification. Broadcast might be a bit confusing because of ActionCable?

also Action Notice ?

**Comment by jonathanhefner** [comment]
_YOU ARE HERE: :round_pushpin: Bikeshed_

Action Notifier has a bit of linguistic overlap with `ActiveSupport::Notifications`.  If we are considering alternative names, what about Action Push?

**Comment by dhh** [comment]
Magnus Hult has very kindly agreed to transfer the gem name for use with this. So the name will be Action Notifier. We will keep the old versions of the gem to ensure no old user is left in a lurch, and we'll start the versioning with Action Notifier 8.0.

**Comment by jbennett** [comment]
This just came up in a conversation on Noticed because I'm setting up an engine for push notifications + noticed. I have a working version here: http://github.com/jbennett/noticed-web_push. 

Takes about 30 seconds to add web push to an app already using Noticed.

**Comment by robinator** [comment]
> But the intention is not to support every possible delivery format for notifications under the sun (like SMS and definitely not email, we have Action Mailer for that).

I'm curious what the "Rails Way" of handling SMS is or should be.  Most apps I've built in the last 10 years have needed it and a nice abstraction ala `ActionMailer` or this proposal would be appreciated.  Fwiw we are currently using `ActiveDelivery` for this.

**Comment by jbennett** [comment]
With the current PWA files solution you are serving up a service_worker.js template. This pulls it out of the normal JS ecosystem used in the rest of the app. Have you considered doing something like having the controller look into the asset pipeline for a service worker entry point and serving that directly?

I've been working on some PWA/WebPush stuff and took that approach so that I could mount the asset from the root and keep everything else in the normal asset pipeline: https://github.com/jbennett/noticed-web_push/blob/132116a5c22bbf491c5e6d09f621f0a65c49ac6d/app/controllers/noticed/web_push/pwa_controller.rb#L4C18-L4C18

**Comment by sandstrom** [comment]
There is a Ruby gem for connection to APNs (Apple Push Notification service):

https://github.com/ostinelli/apnotic

(hopefully it's already well known, but mentioning it here in case it isn't)

**Comment by glaszig** [comment]
> I'm curious what the "Rails Way" of handling SMS is or should be. Most apps I've built in the last 10 years have needed it and a nice abstraction ala `ActionMailer` or this proposal would be appreciated. Fwiw we are currently using `ActiveDelivery` for this.

may i suggest [sms_kit](https://rubygems.org/gems/sms_kit)? i’m the author. let me know if you need something changed or added.

**Comment by ibrahima** [comment]
> Action Notifier should also address the need for controlling the PWA badge counter. That's tied to the hip with Web Push, since the only time you get to reliably move the counter is when a notification is delivered.

I agree that this is very important for parity with native apps. Lately for work I've used a couple web apps that only have a PWA for receiving notifications. At least with the app we're currently using, it has gotten quite annoying that the notifications don't automatically get cleared when you read them on a different device. They end up piling up on my phone over time when I've read them all on my computer. I guess that's different but related to updating the counter badge. I was thinking that you could send a web push for the sake of marking a notification as read, in order to close that existing notification, but apparently [Web Push *requires* showing a notification](https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/How_to/Display_badge_on_app_icon#updating_the_badge_in_the_background). But it does look like there's ways to do background syncs with PWAs that could also be used to update the badge counter and dismiss seen notifications (though sadly the periodic background sync API is not supported in Firefox 😭).

It would be great if Rails/Action Notifier can help with managing the state of notifications across devices, though I don't know how much or how little of a persistence component this is going to have. The notification data schema needed could vary from app to app so it may not make sense to mandate a specific data model. I have been playing with the noticed gem in the last few days and it does have a data model that comes with it, which seems pretty good, but not sure if that's something Action Notifier would want to prescribe. But either way, I'm very excited to see this develop!

**Comment by adenta** [comment]
Where is development of this feature being tracked?

**Comment by adenta** [comment]
@dhh did this get abandoned? in general I can't figure out how to follow along with Rails 8.1, lots of exciting developments I want to track

**Comment by glaszig** [comment]
@adenta those things usually get developed inside 37signals and open sourced at strategic point in time, i.e. around major milestone releases.

**Comment by arpu** [comment]
is there any roadmap or timeline?

**Comment by quantuminformation** [comment]
#1 "rails is not django" ><

**Comment by joemasilotti** [comment]
This can probably be closed now!

https://github.com/rails/action_push_native

**Comment by brunoprietog** [comment]
That's just mobile. Next comes Action Push Web, and probably something to govern them all.
